README
the latest stable pbs_python interface is available from:
    ftp://ftp.surfsara.nl/pub/outgoing/pbs_python.tar.gz

Information, documentation and reporting bugs for the package:
    https://oss.trac.surfsara.nl/oss/pbs_python

bug resport:
    pbs_python@surfsara.nl

=============== Brief description =============================
Pbs_python interface is a wrapper class for the OPENPBS/TORQUE 
C LIB API.  Now you can write utilities/extensions in Python 
instead of C.  

The Python wrapper class is tested on:

  - LISA cluster debian wheezy, torque 2.5.13
  - LTC test cluster debian wheezy, torque 5.0.1
  - LTC test cluster centos 6, torque 5.0.1
  - It should run on different OSes.

-- Compile and install the package: (Require python 2.1 and higher)

  As for python version 2.1 and higher the procedure is:
    ./configure
    make
    make install

  Debian package:
    debian/rules binary

--- Testing the package:

The test programs are include as a reference how to use the pbs
python module. You have to edit some test programs to reflect
your PBS/Torque installation.

pbsmon.py        - ascii xpbsmon
rack_pbsmon.py   - ascii xpbsmon by rack layout
pbsnodes-a.py    - pbsnodes -a
pbs_version.py   - print server version
set_property.py  - set some node properties
resmom_info.py   - queries the pbs_mom daemon on the nodes
logpbs.py	 - Shows the usage of the PBS logging routines
new_interface.py - Example how to use PBSQuery module
PBSQuery.py      - python <install_path>/PBSQuery.py (has builtin demo)

sara_nodes.py    - We use this program to set the nodes offline/online. when 
                   there are no command line arguments. It will list the 
                   nodes that are down/oflline. For more info see:
                    - https://subtrac.sara.nl/oss/pbs_python/wiki/TorqueExamples

pbs_jobmonitor   - Shows the output of top -u user on the node:
                     * one cycle of top
                     * user: the user the job belongs to

pbs_joblogin     - logs in to the node as the user who invokes this script 
                   (os.getenv('USER'))

For more info about SARA see:
    http://www.surfsara.nl/index_eng.html

--
Bas van der Vlies
| Operations, Support & Development | SURFsara | Science Park 140 | 1098 XG  Amsterdam
| T +31 (0) 20 800 1300  | bas.vandervlies@surfsara.nl | www.surfsara.nl |


CHANGES
=========== 4.6.6
 * PBSQuery do not parse simple values. else rendering of offline not
   goes wrong.
   Author: Bas van der Vlies

=========== 4.6.4
 * Use system Python instead of environment Python.
   Author: Martijn Kruiten

=========== 4.6.3
 * Added basic compatibilty with Python 3.x.
   Author: Martijn Kruiten

=========== 4.6.1
 * Patched sara_nodes to work with the pbs and PBSQuery modules
   Author: Dennis Stam

=========== 4.6.0
 * Added support for torque version 5.x. Note: the rm interface does not work for me on debian wheezy.
   Author: Bas van der Vlies

 * PBSQuery had been update to new cpuid range for torque 5
   Author: stijnd : deweirdt add ugent dot be 
   Applied by : Bas van der Vlies

=========== 4.5.0
 * new_rack_pbsmon.py: Added option -j/--job. It will show the nodes that are used by the specified job
 Author: Dennis Stam

 * JobscriptParser.py: Added a new usefull script which parses the jobscript and returns the attropl as datastructure
 Author: Dennis Stam
  

=========== 4.4.1
 * new_rack_pbsmon.py: Added option -p/--property to filter on torque node properties. It will show
   the nodes that meet the property request with a capital M on a non-color terminal and in green on
   color terminals.
 Requested by: Willem Vermin and Wim Rijks
 Author: Dennis Stam

 * Fixed docdir bug in configure.in and Makefile.in, closes #39
 Author: Dennis Stam

=========== 4.4.0
 * Drop support for versions older then 2.4
 Author: Bas van der Vlies

 * Added support for torque version >= 4.2
 Author: Bas van der Vlies

 * rewrote setup.py.in
 Inspired by: stijn dot deweirdt at ugent dot be
 Author: Bas van der Vlies

 * docdir is now configurable, closes #37
 Author: jens dot timmerman add ugent dot be
 Applied by: Bas van der Vlies

 * The swig development environment to generate the pbs_python package
   is now available at:
    - svn co https://oss.trac.surfsara.nl/pbs_python/svn/devel

=========== 4.3.7
 * pbs_jobmonitor/pbs_joblogin ported to torque 2.5 to fix jobarrays
  Author: Willem Vermin (SURFsara)
  Applied by: Bas van der Vlies
  
=========== 4.3.6
 * Modified sara_nodes.py and new_rack_pbsmon.py so:
   - We using argparse instead of OptionParser
   - Changed the print statement to an print function which will work with Python 2.5+
   - Removed module PBSAdvancedParser.py
   Author: Dennis Stam

=========== 4.3.5
 * PBSQuery bugfix, parsing of node status line with multiple EVENT: settings
   was wrong
   Author: Bas van der Vlies

 * Added an example how to make use of the EVENT status message:
    - examples/pbs_script.sh
   Author: Bas van der Vlies

 * implemented __nonzero__ for PBS dictionary, closes #33, eg:
    status = getjob(id)
    if status:
        do something
    else:
        no job
  Author: Bas van der Vlies

 * PBSQuery bugfix. getjob() can now handle short jobname, eg: 
    * p.getjob('6073545') 

   Author: Willem Vermin (SARA) & Bas van der Vlies 

 * added an example: pbs_jobmonitor shows the output of top 
   -u user on the node
    - one cycle of top
    - user: the user the job belongs to

   renamed as pbs_joblogin logs in to the node as the user 
   who invokes this script (os.getenv('USER'))

   Author: Willem Vermin (SARA)
  
=========== 4.3.3
 * New generated files for pbs_wrap.c and pbs.py to support python 3.X versions
   Reported by: Steve Traylen <steve dot traylen add cern dot ch>
   Author: Bas van der Vlies 

 * Fixed AdvancedParser when using 01-12 range the zero was not appended.
   Reported by: Ramon Bastiaans
   Author: Dennis Stam

 * examples/sara_nodes.py: Catch the PBSQuery error if we can not make a 
   connection with the batch server and exit the program.
   Author: Bas van der Vlies 

 * Remove the debian dependency on AdvancedParser. We now have
   our own PBSAdvanceParser so it does not conflict with other SARA
   packages.
   Reported by: Ramon Bastiaans (SARA)
   Fixed by: Bas van der Vlies

 * Make pbs_torque only dependend on libtorque instead of torque and adjusted 
   the maintainer of the package, closes #30
   Reported by: Guillermo Marcus  <guillermo dot marcus add@gmail dot com>
   Fixed by: Bas van der Vlies
   
=========== 4.3.0
 * new_rack_pbsmon.py: For backward compatibility added -a/--all option. 
   Has the same functionality as -s/--summary
   Author: Bas van der Vlies

 * Added some license info to debian package and include GPL and LGPL 3.0
   files.
   Suggested by: W. Trevor King <wking at drexel dot edu>

 * PBSQuery.py: Added support for new feature in node status field:
    * message=EVENT:<key=value:>+
    * message=ERROR <text>

  Added the following syntax for node:
  	* node['error']
    * node['event']

  eg: node: n1
    state = free
    np = 8
    properties = ib,switch1,highmem
    ntype = cluster
    jobs = 0/567403.sara.nl, 1/567403.sara.nl
    status = ..,loadave=0.00,message=EVENT:sample.time=1288864220.003:cputotals.user=0:iconnect.pktout=0,netload=3487600394,state=free,...

  example code:
  	p = PBSQuery()
    node = p.getnode(n1)
    print node.event['sample.time']
    print node['event']['sample.time']

   Requested by: Roy Dragseth <roy dot dragseth add uit dotno>
   Author: Bas van der Vlies

=========== 4.2.0
 * Updated LICENSE file to LGPLV3. Thanks to Justin Bronder 
   (jsbronder add gentoo dot org) for the information and
   explanation.

 * examples/pbsmon.py. Patched so it also works for torque 2.5
   Author: Stijn De Weirdt stijn dot deweirdt at ugent dot be
   Applied by: Bas van der Vlies

 * setup.py.in changed determine which torque version is installed
   so we also support snapshots.
   Author: Mark Roberts ( mark at gingergeeks dot co dot uk)
   Applied by: Bas van der Vlies

 * PBSQuery:
   - get_nodes function for jobs now works for old and new datastructure and
     gives correct results if unique=True
     Reported by + idea for patch: Shunjie Lau (eijnuhs add gmail dot com)
     Author: Bas van der Vlies


setup_no_autotools.py
#!/usr/bin/env python
#
# $Id: setup.py 434 2005-11-04 15:02:07Z bas $
#
# set ts=4
#

"""
Based on setup.py from https://github.com/ehiggs/pbs-python
@author: Ewan Higgs
"""

import sys
import os
import glob
import subprocess
import re

from distutils.core import setup, Extension
from distutils.version import LooseVersion

SOURCE_DIR = 'src'

def pbs_config():
    return 'pbs-config'

def pbs_library_compile_line(pbs_conf):
    out, err = subprocess.Popen([pbs_conf,  '--libs'], stdout=subprocess.PIPE).communicate()
    return out.rstrip()

def pbs_library(pbs_conf):
    out, err = subprocess.Popen([pbs_conf, '--libs'], stdout=subprocess.PIPE).communicate()
    stripped_prefix = out[2:].rstrip() #'-l'
    return stripped_prefix

def pbs_library_dir(pbs_conf):
    out, err = subprocess.Popen([pbs_conf, '--libdir'], stdout=subprocess.PIPE).communicate()
    return out.rstrip()

def pbs_version(pbs_conf):
    out, err = subprocess.Popen([pbs_conf, '--version'], stdout=subprocess.PIPE).communicate()
    return out.rstrip()

def my_version():
    """Extract current version from configure.in"""
    conf_in = os.path.join(os.path.dirname(__file__), 'configure.in')
    txt = open(conf_in).read()
    # patttern is AC_INIT([pbs-python], [4.6.1])
    reg =  re.compile(r'^AC_INIT\(\[pbs-python\],\s*\[(\d+\.\d+(?:\.\d+)?)\]\)', re.M)
    found = reg.search(txt)
    if found:
        return found.group(1)
    else:
        raise Exception("Unable to determine pbs_python version from configure.in at %s" % conf_in)

if __name__ == '__main__':
    pbs_conf = pbs_config()
    PBS_LIB_COMPILE_LINE=pbs_library_compile_line(pbs_conf)
    PBS_LIB_DIR=pbs_library_dir(pbs_conf)
    PBS_LIBS = pbs_library(pbs_conf)

    PBS_VERSION = pbs_version(pbs_conf)
    VERSION = my_version()

    try:
        os.unlink('src/pbs.py')
    except OSError:
        pass

    if LooseVersion(PBS_VERSION) >= LooseVersion('5.0'):
        TORQUE_VERSION='TORQUE_5'
        SOURCE_FILE='src/5.x/pbs_wrap.cxx'
        PBS_PYTHON_INCLUDE_DIR = 'src/5.x'

        if not os.path.exists(os.path.join(PBS_PYTHON_INCLUDE_DIR,'log.h')):
            print 'Failed to find log.h in include dir %s. (Set include dir via PBS_PYTHON_INCLUDEDIR variable)'%inc
            sys.exit(2)

        for fn in glob.glob('*.h'):
            os.remove(fn)

        os.symlink('5.x/pbs.py', 'src/pbs.py')

    elif LooseVersion(PBS_VERSION) >= LooseVersion('4.2'):
        TORQUE_VERSION='TORQUE_4'
        SOURCE_FILE='src/C++/pbs_wrap.cxx'
        PBS_PYTHON_INCLUDE_DIR = 'src/C++'

        os.symlink('C++/pbs.py', 'src/pbs.py')

    elif LooseVersion(PBS_VERSION) >= LooseVersion('2.4'):
        TORQUE_VERSION='TORQUE_2'
        SOURCE_FILE='src/C/pbs_wrap.c'
        PBS_PYTHON_INCLUDE_DIR = 'src/C'
        # Older pbs-config yielded some link line info instead of just the libraries
        PBS_LIBS = 'torque'

        os.symlink('C/pbs.py', 'src/pbs.py')

    else:
        print "Version: %s is not supported" %(PBS_VERSION)
        sys.exit(1)

    INCLUDE_DIR = '/usr/include/torque'

    setup (
        name = 'pbs_python',
        version = VERSION,
        description = 'openpbs/torque python interface',
        license = 'LGPLV3',
        author = 'Bas van der Vlies',
        author_email = 'bas.vandervlies@surfsara.nl',
        url = 'http://oss.trac.surfsara.nl/pbs_python',


        extra_path = 'pbs',
            package_dir = { '' : SOURCE_DIR },
            py_modules = [ 'pbs',  'PBSQuery' ],

        ext_modules = [
            Extension( '_pbs', [SOURCE_FILE],
                define_macros =  [ (TORQUE_VERSION, None) ],
                include_dirs = [ INCLUDE_DIR, PBS_PYTHON_INCLUDE_DIR ],
                library_dirs = [ PBS_LIB_DIR ],
                libraries = [ PBS_LIBS ],
                )
        ]
    )
